pkg/operator/rbaccontroller/rbaccontroller.go

```diff
@@ -0,0 +1,106 @@
+	openShiftManagedRBACAnnotation = "openshiftapiserver.rbac.authorization.openshift.io/managed"
+)
+
+// RBACController reconciles RBAC resources created by openshift-apiserver.
+//
+// If openshiftManagedRBACAnnotation is set to "true" for a cluster role binding,
+// RBACController would remove unauthenticated subject from list of subjects.
+//
+// A user can opt-out of this reconciliation by setting the annotation to "false".
+type RBACController struct {
+	eventRecorder            events.Recorder
+	kubeClient               kubernetes.Interface
+	clusterRoleBindingLister listerrbacv1.ClusterRoleBindingLister
+}
+
+func NewRBACController(kubeClient kubernetes.Interface, informerFactory informers.SharedInformerFactory, recorder events.Recorder) factory.Controller {
+	c := &RBACController{
+		kubeClient:               kubeClient,
+		clusterRoleBindingLister: informerFactory.Rbac().V1().ClusterRoleBindings().Lister(),
+		eventRecorder:            recorder,
+	}
+
+	return factory.New().
+		WithInformers(informerFactory.Rbac().V1().ClusterRoleBindings().Informer()).
+		WithSync(c.sync).
+		ToController("RBACController", c.eventRecorder)
+}
+
+// sync ensures that cluster role bindings created by openshift-apiserver do not have the unauthenticated subject.
+func (c *RBACController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	// The named cluster role bindings are created by openshift-apiserver
+	// and in previous releases may have included the unauthenticated subject.
+	clusterRoleBindingNames := []string{"discovery", "basic-users", "system:openshift:discovery",
+		"cluster-status-binding"}
+
+	var errs []error
+
+	for _, clusterRoleBindingName := range clusterRoleBindingNames {
+		clusterRoleBinding, err := c.clusterRoleBindingLister.Get(clusterRoleBindingName)
+		if apierrors.IsNotFound(err) {
+			// Not every cluster role binding is present in every version of OpenShift.
+			// Downgrading the error to avoid log spamming.
+			klog.V(4).Infof("fetching cluster role binding %s failed %v", clusterRoleBindingName, err)
+			continue
+		} else if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		removeUnauthenticatedSubject := (clusterRoleBinding.Annotations[openShiftManagedRBACAnnotation] == "true")
+		if !removeUnauthenticatedSubject {
+			continue
+		}
+
+		updatedCopy := withUnauthenticatedSubjectRemoved(clusterRoleBinding)
+		if updatedCopy == nil {
+			continue
+		}
+
+		_, err = c.kubeClient.RbacV1().ClusterRoleBindings().Update(ctx, updatedCopy, metav1.UpdateOptions{})
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+// withUnauthenticatedSubjectRemoved returns a new copy of given cluster role binding without unauthenticated user in subjects.
+func withUnauthenticatedSubjectRemoved(clusterRoleBinding *rbacv1.ClusterRoleBinding) *rbacv1.ClusterRoleBinding {
+	if clusterRoleBinding == nil {
```

Comment:
I left it as a safe guard so that code does not panic when nil object is accidentally passed by a new caller.

Reply:
(No action required, just food for thought)

There's this principle I really like - YAGNI (you ain't gonna need it). Err on the side of doing less if it isn't required today. It may be that more will never be required, and if it is, you can do it in the future instead of now. 

Another thought - have the caller dereference the pointer. The return value is a copy after all - the function doesn't have to modify it.